File: Python/261__Graph_Valid_Tree.py
class Solution(object):
    def validTree(self, n, edges):
        """
        :type n: int
        :type edges: List[List[int]]
        :rtype: bool
        """
        def findSubset(i):
            if union[i] == -1:
                return i
            else:
                return findSubset(union[i])

        union = [-1] * n
        for edge in edges:
            s1 = findSubset(edge[0])
            s2 = findSubset(edge[1])

            if s1 == s2:
                return False
            else:
                union[s1] = s2
        if len(edges) != n - 1:
            return False
        else:
            return True
            

File: Java/261__Graph_Valid_Tree.java
class Solution {

    static class DSU {
        int[] parent;
        DSU (int size) {
            parent = new int[size];
            for (int i = 0; i < size; i++)
                parent[i] = i;
        }

        public int find(int x) {
            if (x != parent[x])
                parent[x] = find(parent[x]);
            return parent[x];
        }

        public boolean union(int x, int y) {
            int xr = find(x), yr = find(y);
            if (xr == yr) return false;
            parent[xr] = yr;
            return true;
        }
    }

    public boolean validTree(int n, int[][] edges) {
        DSU dsu = new DSU(n);
        for (int[] edge : edges)
            if (!dsu.union(edge[0], edge[1]))
                return false;
        int count = 0;
        for (int i = 0; i < n; i++)
            if (dsu.find(i) == i)
                count++;
        return count == 1;
    }
}
